packages/docusaurus-utils/src/tags.ts

```diff
@@ -159,3 +214,82 @@ export function getTagVisibility<Item>({
     listedItems: items.filter((item) => !isUnlisted(item)),
   };
 }
+
+export function validateFrontMatterTags({
+  tags,
+  source,
+  options,
+}: {
+  tags: NormalizedTag[];
+  source: string;
+  options: Pick<PluginOptions, 'tagsFilePath' | 'onUnknownTags'>;
+}): void {
+  const inlineTags = tags.filter((tag) => tag.inline);
+  if (inlineTags.length > 0 && options.onUnknownTags !== 'ignore') {
+    const uniqueUnknownTags = [...new Set(inlineTags.map((tag) => tag.label))];
+    const tagListString = uniqueUnknownTags.join(', ');
+    // @ts-expect-error: onUnknownTags is not 'ignore'
+    logger.report(options.onUnknownTags)(
+      `Tags [${tagListString}] used in ${source} are not defined in ${options.tagsFilePath}`,
+    );
+  }
+}
+
+export function processFileTagsPath({
+  options,
+  source,
+  frontMatterTags,
+  versionTagsPath,
+  tagsFile,
+}: {
+  options: Pick<PluginOptions, 'tagsFilePath' | 'onUnknownTags'>;
+  source: string;
+  frontMatterTags: FrontMatterTag[] | undefined;
+  versionTagsPath: string;
+  tagsFile: TagsFile | undefined;
+}): NormalizedTag[] {
+  const tags = normalizeTags({
+    versionTagsPath,
+    tagsFile,
+    frontMatterTags: frontMatterTags ?? [],
+  });
+
+  validateFrontMatterTags({
+    tags,
+    source,
+    options,
+  });
+
+  return tags;
+}
+
+export async function getTagsFile<T>(
```

Comment:
Needed to change this to generic type and also pass the validation function as parameter because now in docusaurus utils there is no joi, the function `validateDefinedTags` is in `docusaurus utils validation` and cannot be imported but not sure about the whole thing